### ISSUE TITLE
Passing of TLS parameters

### ISSUE THREAD

**Issue by stefanberger** [flimzy/kivik#252]
It doesn't seem to be possible to pass TLS parameters to New() so that for example client certificates can be passed, among other things. Would it be possible to add a parameter to this function or create a new function that allows passing these parameters through to the driver that then creates http.Client{} with these parameters?

**Comment by flimzy** [comment]
Good question. It may not be very straight forward to do that. Let me look into it.

**Comment by flimzy** [comment]
I have devised a possible solution. I would welcome your feedback, as to whether you think it's a feasible approach.

https://github.com/go-kivik/couchdb/pull/60 adds a new authenticator method which allows providing a custom [http.RoundTripper](https://golang.org/pkg/net/http/#RoundTripper).  This allows you specify an [http.Transport](https://golang.org/pkg/net/http/#Transport), which may provide TLS config, or any of the other connection-level HTTP config (proxies, keep-alives, compression, etc).

To use it, you'll need to use the latest (master) version of the couchdb driver, then in your code, do something like:

```go
client, _ := kivik.New("couchdb", ...)
setXport := couchdb.SetTransport(&http.Transport{
    TLSClientConfig: &tls.Config{ ... },
})
_ := client.Authenticate(setXport)
```

One caveat with this method: If being used in conjunction with BasicAuth, you must perform BasicAuth _after_ setting the custom transport. But I expect that won't affect you, if you're using TLS for auth.

Please let me know if you have any questions.

At the moment, please consider this API experimental.  I may change details based on your feedback, or after pondering it a while.

**Comment by stefanberger** [comment]
Thanks a lot for the quick turn-around.

One of the more trivial things I want to do is to be able to have the client accept "certificates signed by an unknown authority" (e.g., self-signed certs), primarily for testing. But kivik.New() already throws an error when it gets a cert it doesn't know the CA for... So I think it would have to be a parameter to New().

**Comment by flimzy** [comment]
> But kivik.New() already throws an error when it gets a cert it doesn't know the CA for...

Hmmm, yeah, that will definitely require additional thought, then.  As you say, an optional parameter to `New()` may be the way to go.  I'll continue to mull this one over.

**Comment by flimzy** [comment]
One follow-up question:  Are you passing auth credentials to `kivik.New()`?  If you do, that could explain the connection-time error, as it attempts to validate the session immediately.  Perhaps it would be an option to instead try this:

```go
client, _ := kivik.New("couchdb", ...) // Make sure no credentials are included
setXport := couchdb.SetTransport(&http.Transport{
    TLSClientConfig: &tls.Config{ ... },
})
_ := client.Authenticate(setXport) // Set up TLS config
err := client.Authenticate(&chttp.BasicAuth{Username: "xxx", Password: "xxx"}) // Basic Auth
// or
err := client.Authenticate(&chttp.CookieAuth{Username: "xxx", Password: "xxx"}) // Cookie Auth
```

Note that these `chttp.*Auth` types are effectively undocumented, which I need to remedy.

**Comment by stefanberger** [comment]
You're right. I have the username and password encoded in the URL of course and that triggers connection establishment. I will try your suggestion later.

**Comment by stefanberger** [comment]
So this works for me.
When I debugged this I think I saw it creating multiple connections to the server, not just one. Not sure why this would be, but it works...

**Comment by flimzy** [comment]
> So this works for me.

Glad to hear it!

> When I debugged this I think I saw it creating multiple connections to the server, not just one. Not sure why this would be, but it works...

If you can provide additional details, such as logs from CouchDB showing the multiple connections, that could be helpful.  Otherwise, I may see if I can debug it myself when time permits.

**Comment by stefanberger** [comment]
I have approximately a function like the one below in my code. The CouchDB is running in a docker container that is accessible for https via port 6984. (The docker-proxy that proxies this port for the container creates a 2nd connection for every connection created.) I am running `ssldump -i any port 6984` and it seems to go through ssl connection establishment twice (4 connection establishments visible in ssl but due to proxying have to half that).

```
	// remove username and password from URL to avoid authentication step
	// before we can set the TLS transport
	_host, err := url.Parse(host)
	_host.User = nil

	client, err := kivik.New(context.TODO(), "couch", _host.String())
	if err != nil {
		return nil, err
	}

	setXport := couch.SetTransport(&http.Transport{
		TLSClientConfig: &tls.Config{}
	})

	if err = client.Authenticate(context.TODO(), setXport); err != nil {
		return nil, err
	}

	if username != "" {
		var basicAuth chttp.Authenticator = &chttp.BasicAuth {
			Username: username,
			Password: password,
		}
		if err = client.Authenticate(context.TODO(), basicAuth); err != nil {
			return nil, err
		}
	}
	os.Exit(0)
```

**Comment by stefanberger** [comment]
@flimzy I am working on a CouchDB driver for the Docker [Notary](https://github.com/theupdateframework/notary). For that I would need the HTTPs support. Are you planning on merging the https support soon? Would you mind me sending them a pull request noting that the patch series for CouchDB is work-in-progress, also because the Kivik HTTPs support is not there, yet. I would just be interested in getting feedback from them about all the other code.

**Comment by flimzy** [comment]
That support was merged and rolled out in the [v1.7.0](https://github.com/go-kivik/couchdb/tree/v1.7.0) release just a few days ago.

**Comment by stefanberger** [comment]
Even better. I guess we can close this issue here.

**Comment by flimzy** [comment]
Cool.  I'll open a new issue to track the double connections then.

**Comment by stefanberger** [comment]
Ok.
